alphadia/workflow/optimizer.py

```diff
@@ -0,0 +1,147 @@
+        self.reporter = reporting.LogBackend() if reporter is None else reporter
+
+    def check_stopping_conditions(self):
+        raise NotImplementedError(
+            f"check_stopping_conditions() not implemented for {self.__class__.__name__}"
+        )
+
+    def update(self):
+        raise NotImplementedError(
+            f"update() not implemented for {self.__class__.__name__}"
+        )
+
+    def initiate(self):
+        raise NotImplementedError(f"initiate() not implemented for {self.__class__}")
+
+
+class RTOptimizer(BaseOptimizer):
+    pass
+
+
+class MS2Optimizer(BaseOptimizer):
+    def __init__(self, initial_tolerance: float):
+        super().__init__(initial_tolerance)
+        self.estimator = "mz"
+        self.df = "fragment"
+        self.ids = []
+
+    @check_if_optimum_found
+    def check_stopping_conditions(self):
+        """Optimization should stop if continued narrowing of the MS2 tolerance is not improving the number of precursor identifications.
+        This function checks if the previous rounds of optimization have led to a meaningful improvement in the number of identifications.
+        If so, it continues optimization and appends the proposed new tolerance to the list of tolerances. If not, it stops optimization and sets the optimal tolerance attribute.
+
+        Notes
+        -----
+            Because the check for an increase in identifications requires two previous rounds, the function will also initialize for another round of optimization if the round variable is less than 2.
+
+            Once the optimal tolerance is set, the optimizer will no longer check for improvements.
+
+        Returns
+        -------
+        stop_optimizaton: bool
+            True if there has been no recent improvement in the number of identifications and optimization should stop. False otherwise.
+
+        """
+
+        if (
+            self.round < 2
+            or self.ids[-1] > 1.1 * self.ids[-2]
+            or self.ids[-1] > 1.1 * self.ids[-3]
+        ):
+            self.reporter.log_string(
+                f"❌ MS2: optimization incomplete in round {self.round}. Will search with tolerance {self.proposed_new_tolerance}.",
+                verbosity="progress",
+            )
+            stop_optimizaton = False
+
+        else:
+            self.optimal_tolerance = self.tolerances[np.argmax(self.ids)]
+            self.end_of_optimization_message()
+            stop_optimizaton = True
+
+        return stop_optimizaton
+
+    @check_if_optimum_found
+    def initiate(self):
+        """If the optimization continues (which will use the proposed new tolerance), the round variable should be incremented and the proposed new tolerance should be added to the list of tolerances."""
+        self.tolerances.append(self.proposed_new_tolerance)
+        self.round += 1
+
+    @check_if_optimum_found
+    def update(self, precursors_df: pd.DataFrame, proposed_new_tolerance: float):
```

Comment:
I think this is the place where we want to calcualte the `proposed_new tolerance`.

Reply:
I agree this would be preferable but at the moment the means of calculating a new tolerance requires the calibration_manager and I wasn't sure we wanted to create a dependency of the optimizer object on the calibration_manager. Do you think I should do this?

Reply:
Yes, that's fine. This class doesn't need to be general.